internal/controller/config/config.go

```diff
@@ -0,0 +1,70 @@
+package config
+
+import (
+	"fmt"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GlobalConfig is the configuration for the controllers, it is
+// supposed to be populated from the configmap attached to the
+// controller manager.
+type GlobalConfig struct {
+	ClusterName     string                    `yaml:"clusterName" json:"clusterName"`
+	TimeIntervalSec uint                      `yaml:"timeIntervalSec" json:"timeIntervalSec"`
+	RepoCredentials map[string]*GitCredential `json:"repoCredentials" yaml:"repoCredentials"`
+}
+
+type GitCredential struct {
+	HTTPCredential *HTTPCredential
+	SSHCredential  *SSHCredential
+	TLS            *TLS
+}
+
+type HTTPCredential struct {
+	Username string
+	Password *corev1.SecretKeySelector
+}
+type SSHCredential struct {
+	SSHKey *corev1.SecretKeySelector
+}
+
+type TLS struct {
+	// +optional
+	InsecureSkipVerify bool `json:"insecureSkipVerify,omitempty" protobuf:"bytes,1,opt,name=insecureSkipVerify"`
+	// CACertSecret refers to the secret that contains the CA cert
+	// +optional
+	CACertSecret *corev1.SecretKeySelector `json:"caCertSecret,omitempty" protobuf:"bytes,2,opt,name=caCertSecret"`
+	// CertSecret refers to the secret that contains the cert
+	// +optional
+	CertSecret *corev1.SecretKeySelector `json:"clientCertSecret,omitempty" protobuf:"bytes,3,opt,name=certSecret"`
+	// KeySecret refers to the secret that contains the key
+	// +optional
+	KeySecret *corev1.SecretKeySelector `json:"clientKeySecret,omitempty" protobuf:"bytes,4,opt,name=keySecret"`
+}
+
+func LoadConfig(onErrorReloading func(error)) (*GlobalConfig, error) {
+	v := viper.New()
+	v.SetConfigName("config")
+	v.SetConfigType("yaml")
+	v.AddConfigPath("/etc/numaplane")
+	err := v.ReadInConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load configuration file. %w", err)
+	}
+	r := &GlobalConfig{}
+	err = v.Unmarshal(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed unmarshal configuration file. %w", err)
+	}
+	v.WatchConfig()
+	v.OnConfigChange(func(e fsnotify.Event) {
+		err = v.Unmarshal(r)
+		if err != nil {
+			onErrorReloading(err)
+		}
+	})
+	return r, nil
+}
```

Comment:
I'm thinking that the way to deal with [this](https://github.com/numaproj-labs/numaplane/pull/80#issuecomment-1934681589) concurrency issue may be to create a struct called `ConfigManager` which holds the official copy of the `GlobalConfig`. At any time somebody can call `func (cm *ConfigManager) getConfig() GlobalConfig`, which you'll notice returns a copy of the `GlobalConfig` which is guaranteed to be atomically correct. Any user should call this once to get the `Config` and use it to get the various fields from it.

`ConfigManager` can have a `GlobalConfig` as a member, which is what it would return when `GetConfig()` is called. 
This `LoadConfig()` function could be a member of the `ConfigManager` struct. It should no longer return a `*GlobalConfig`. Instead it will simply set its own `GlobalConfig` member. 
`v.OnConfigChange()` will set the member as well. The `v.OnConfigChange()` code block and the `GetConfig()` method will need to both use the same mutex

Reply:
ok got it .I will make the changes

Reply:
So, to summarize, something like this:
```
type ConfigManager struct {
   config GlobalConfig
   lock sync.RWMutex
}
func (cm *ConfigManager) GetConfig() GlobalConfig {
   lock.RLock()
   defer lock.RUnlock()
   return config
}
func (cm *ConfigManager) LoadConfig() error {
    // add logic to set 'config' here
    ....
    v.OnConfigChange(func(e fsnotify.Event) {
                lock.WLock()
                defer lock.WUnlock()
		// add logic to set 'config' here
	})
}

```

Reply:
This is right. https://github.com/numaproj-labs/log-ingestor/blob/main/pkg/ingestor/ingestor.go#L168

Reply:
Oh, I guess you already thought about this issue. :) In our case, I also don't want the user to hold a reference to the config either: otherwise they might for example get the Username field and then the Password field, and in between the reference got changed so it's a mismatch.